File: src/priorityq.py
"""Class that initializes an instance of a priority queue."""


class Priority(object):
    """Class to create priority queue."""

    def __init__(self):
        """Initialize instance of self."""
        self._counter = 0
        self.priorities = []
        self.head = None
        self.tail = None

    def insert(self, val, priority=0):
        """Add value to end of queue."""
        if priority < 0:
            raise ValueError('Lowest priority is 0.')
        self.priorities.append(priority)
        if self.head is None:
            curr = Node(val, priority)
            self.head = curr
            self.tail = self.head
            self._counter += 1
        else:
            curr = Node(val, priority, next_node=None, prev_node=self.tail)
            self.tail.next_node = curr
            self.tail = curr
            self._counter += 1

    def pop(self):
        """Remove highest priority value from queue."""
        if self.head is None:
            raise IndexError('Input priority not in Priority Queue.')
        max_priority = max(self.priorities)
        current_node = self.head
        while current_node is not None:
            print(current_node.priority)
            if current_node.priority == max_priority:
                val = current_node.data
                if current_node is self.head:
                    self.head = None
                    self._counter -= 1
                    return val
                if current_node.prev_node is None:
                    current_node.next_node.prev_node = None
                    self.head = None
                    self._counter -= 1
                    return val
                if current_node.next_node is None:
                    current_node.prev_node.next_node = None
                    self.tail = None
                    self._counter -= 1
                    return val
                current_node.prev_node.next_node = current_node.next_node
                current_node.next_node.prev_node = current_node.prev_node
                self._counter -= 1
                self.priorities.remove(max_priority)
                return val
            current_node = current_node.next_node

    def peek(self):
        """Return highest priority node in queue."""
        if self.head is None:
            raise IndexError('Priority Queue is empty.')
        max_priority = max(self.priorities)
        current_node = self.head
        while current_node is not None:
            if current_node.priority == max_priority:
                val = current_node.data
                return val
            current_node = current_node.next_node


class Node(object):
    """Double List Node class."""

    def __init__(self, val, priority, next_node=None, prev_node=None):
        """Initialize Node class instance."""
        self.priority = priority
        self.data = val
        self.next_node = next_node
        self.prev_node = prev_node

File: js/priority-q.js
'use strict';

class Node {
  constructor(val, priority, nextNode=null, prevNode=null) {
    this.priority = priority;
    this.data = val;
    this.nextNode = nextNode;
    this.prevNode = prevNode;
  }
}

class Priority {
  constructor() {
    this._counter = 0;
    this.priorities = [];
    this.head = null;
    this.tail = null;
  }

  insert(val, priority=0) {
    if (priority < 0) {
      throw new Error('Lowest priority is 0.')
    }
    this.priorities.push(priority);
    if (!this.head) {
      let currentNode = new Node(val, priority);
      this.head = currentNode;
      this.tail = this.head;
      this._counter ++;
    } else {
      let currentNode = new Node(val, priority, null, this.tail);
      this.tail.nextNode = currentNode;
      this.tail = currentNode;
      this._counter ++;
    }
  }

  pop() {
    if (!this.head) {
      throw new Error('Input priority not in Priority Queue.');
    }
    let maxPriority = this.priorities.reduce(function(a, b) {
      return Math.max(a, b);
    });
    let currentNode = this.head;
    while (currentNode) {
      if (currentNode.priority == maxPriority) {
        let val = currentNode.data;
        if (currentNode === this.head && this._counter == 1) {
          this.head = null;
          this.tail = null;
          this._counter --;
          return val;
        } else if (!currentNode.prevNode) {
          currentNode.nextNode.prevNode = null;
          this.head = null;
          this._counter --;
          return val;
        } else if (!currentNode.nextNode) {
          currentNode.prevNode.nextNode = null;
          this.tail = null;
          this._counter --;
          return val;
        }
        currentNode.prevNode.nextNode = currentNode.nextNode;
        currentNode.nextNode.prevNode = currentNode.prevNode;
        this._counter --;
        this.priorities.splice(this.priorities.indexOf(val), 1);
        return val;
      }
      currentNode = currentNode.nextNode;
    }
  }

  peek() {
    if (!this.head) {
      throw new Error('Priority Queue is empty.');
    }
    let maxPriority = this.priorities.reduce(function(a, b) {
      return Math.max(a, b);
    });
    let currentNode = this.head;
    while (currentNode) {
      if (currentNode.priority == maxPriority) {
        let val = currentNode.data;
        return val;
      }
      currentNode = currentNode.nextNode;
    }
  }

}

module.exports = Priority
